File: tests/comparison.test.js
// comparison.js — Dialekt-Vergleich nach gleicher Bedeutung.

import { describe, it } from 'node:test';
import assert from 'node:assert/strict';

import { buildComparison, filterByKategorie } from '../js/util/comparison.js';

describe('buildComparison', () => {
  it('liefert Gruppen mit head/items/kategorie', () => {
    const groups = buildComparison({ minSize: 2 });
    assert.ok(groups.length > 0);
    for (const g of groups.slice(0, 5)) {
      assert.ok(typeof g.head === 'string');
      assert.ok(Array.isArray(g.items));
      assert.ok(typeof g.kategorie === 'string');
      assert.ok(g.items.length >= 2);
    }
  });

  it('default minSize=2 — alle Gruppen ≥2 Dialekte', () => {
    const groups = buildComparison();
    assert.ok(groups.every(g => g.items.length >= 2));
  });

  it('mit minSize=5 nur größere Gruppen', () => {
    const groups = buildComparison({ minSize: 5 });
    assert.ok(groups.every(g => g.items.length >= 5));
  });

  it('sortiert nach Größe (descending)', () => {
    const groups = buildComparison({ minSize: 3 });
    for (let i = 1; i < groups.length; i++) {
      assert.ok(groups[i - 1].items.length >= groups[i].items.length);
    }
  });

  it('je Gruppe nur 1 Ausdruck pro Dialekt', () => {
    const groups = buildComparison({ minSize: 2 });
    for (const g of groups.slice(0, 20)) {
      const dialektIds = g.items.map(i => i.dialektId);
      assert.equal(new Set(dialektIds).size, dialektIds.length);
    }
  });
});

describe('filterByKategorie', () => {
  const groups = buildComparison({ minSize: 2 });

  it('"all" liefert alle Gruppen', () => {
    assert.equal(filterByKategorie(groups, 'all').length, groups.length);
  });

  it('null/undefined liefert alle Gruppen', () => {
    assert.equal(filterByKategorie(groups, null).length, groups.length);
    assert.equal(filterByKategorie(groups, undefined).length, groups.length);
  });

  it('Filter nach spezifischer Kategorie', () => {
    const greetings = filterByKategorie(groups, 'begruessung');
    assert.ok(greetings.every(g => g.kategorie === 'begruessung'));
  });
});

File: mobile/test/comparison_test.dart
// Tests für die Vergleichs-Gruppierung — Port von tests/comparison.test.js.

import 'package:flutter_test/flutter_test.dart';

import 'package:dialekto/data/models.dart';
import 'package:dialekto/util/comparison.dart';

DialektAusdruck pair(
  String dId,
  String hd, {
  String kat = 'alltag',
  String ausdruck = 'x',
}) {
  final d = Dialekt.fromJson({'id': dId, 'name': dId, 'ausdruecke': const []});
  final a = Ausdruck.fromJson({
    'id': '$dId-1',
    'ausdruck': ausdruck,
    'hochdeutsch': hd,
    'kategorie': kat,
  });
  return (dialekt: d, ausdruck: a);
}

void main() {
  test('gruppiert gleiche Bedeutung über Dialekte', () {
    final groups = buildComparison([
      pair('a', 'Junge', ausdruck: 'Bub'),
      pair('b', 'Junge', ausdruck: 'Knab'),
    ]);
    expect(groups.length, 1);
    expect(groups.first.size, 2);
    expect(groups.first.head, 'Junge');
  });

  test('nur ein Ausdruck pro Dialekt pro Gruppe', () {
    final groups = buildComparison([
      pair('a', 'Junge', ausdruck: 'Bub'),
      pair('a', 'Junge', ausdruck: 'Bua'), // selber Dialekt → ignoriert
      pair('b', 'Junge', ausdruck: 'Knab'),
    ]);
    expect(groups.length, 1);
    expect(groups.first.size, 2);
  });

  test('minSize filtert Einzelgruppen', () {
    final groups = buildComparison([
      pair('a', 'Junge'),
      pair('b', 'Junge'),
      pair('a', 'Mädchen'), // nur 1 Dialekt
    ]);
    expect(groups.length, 1);
    expect(groups.first.head, 'Junge');
  });

  test('Token-Overlap merged Varianten (Umstellung), aber nicht schwache', () {
    final merged = buildComparison([
      pair('a', 'der Junge'),
      pair('b', 'Junge der'),
    ]);
    expect(merged.length, 1);
    expect(merged.first.size, 2);

    final split = buildComparison([
      pair('a', 'kleiner Junge'),
      pair('b', 'Junge'),
    ], minSize: 1);
    expect(split.length, 2);
  });

  test('Kopf wird auf Trenner gesplittet', () {
    final g = buildComparison([
      pair('a', 'Kartoffel / Erdapfel'),
      pair('b', 'Kartoffel'),
    ]);
    expect(g.length, 1);
    expect(g.first.head, 'Kartoffel');
  });

  test('sortiert nach Gruppengröße absteigend', () {
    final g = buildComparison([
      pair('a', 'Junge'),
      pair('b', 'Junge'),
      pair('c', 'Junge'),
      pair('a', 'Haus'),
      pair('b', 'Haus'),
    ]);
    expect(g.first.head, 'Junge');
    expect(g.first.size, 3);
    expect(g.last.size, 2);
  });

  test('filterByKategorie', () {
    final g = buildComparison([
      pair('a', 'Junge', kat: 'menschen'),
      pair('b', 'Junge', kat: 'menschen'),
      pair('a', 'Haus', kat: 'wohnen'),
      pair('b', 'Haus', kat: 'wohnen'),
    ]);
    expect(filterByKategorie(g, 'all').length, 2);
    final menschen = filterByKategorie(g, 'menschen');
    expect(menschen.length, 1);
    expect(menschen.first.head, 'Junge');
  });
}
